File: scala/Runner.scala
package tqlmate

import java.nio.file.{Files, Path}
import java.time.{Instant, ZoneOffset}
import java.time.format.DateTimeFormatter
import java.time.temporal.ChronoUnit

import scala.concurrent.duration.*
import scala.util.Using

import com.typedb.driver.TypeDB
import com.typedb.driver.api.{Credentials, Driver, DriverOptions, DriverTlsConfig}

final case class RunnerOpts(
  url: TypeDbUrl,
  migrationsDir: Path,
  schemaFile: Path,
  strict: Boolean,
  verbose: Boolean,
  waitTimeout: Option[FiniteDuration]
)

final class Runner(opts: RunnerOpts):
  private var driverSlot: Option[Driver] = None

  private def driver(): Result[Driver] =
    driverSlot match
      case Some(d) => Right(d)
      case None =>
        val opened =
          opts.waitTimeout match
            case Some(t) => Runner.waitForServer(opts.url, t, opts.verbose).flatMap(_ => Runner.openDriver(opts.url))
            case None    => Runner.openDriver(opts.url)
        opened.map: d =>
          driverSlot = Some(d)
          d

  def close(): Unit =
    driverSlot.foreach(_.close())
    driverSlot = None

  def create(): Result[Unit] =
    val name = opts.url.database
    driver().flatMap: d =>
      Runner.edge:
        if d.databases().contains(name) then
          if opts.verbose then System.err.println(s"database already exists: $name")
        else
          d.databases().create(name)
          println(s"Created: $name")

  def drop(): Result[Unit] =
    val name = opts.url.database
    driver().flatMap: d =>
      Runner.edge:
        if !d.databases().contains(name) then
          if opts.verbose then System.err.println(s"database does not exist: $name")
        else
          d.databases().get(name).delete()
          println(s"Dropped: $name")

  def newMigration(name: String): Result[Path] =
    Runner.edge:
      Files.createDirectories(opts.migrationsDir)
      val path = Migration.newMigrationPath(opts.migrationsDir, name)
      Files.writeString(path, Migration.template)
      println(s"Created: $path")
      path

  def migrate(): Result[Unit] =
    val db = opts.url.database
    for
      d <- driver()
      exists <- Runner.edge(d.databases().contains(db))
      _ <-
        if !exists then Left(Error.msg(s"database does not exist: $db (run create or up)"))
        else Right(())
      _ <- Ledger.ensure(d, db)
      files <- Migration.listMigrationFiles(opts.migrationsDir)
      applied <- Ledger.appliedVersions(d, db)
      _ <- if opts.strict then Migration.checkStrictOrder(files, applied) else Right(())
      pending = files.filterNot(f => applied.contains(f.version))
      _ <-
        if pending.isEmpty then
          if opts.verbose then System.err.println("Migrations: nothing to apply")
          Right(())
        else
          pending.foldLeft[Result[Unit]](Right(())): (acc, m) =>
            acc.flatMap: _ =>
              Runner.applyUp(d, db, m.version, m.up, opts.verbose).map: _ =>
                println(s"Applied: ${m.label}")
    yield ()

  def rollback(): Result[Unit] =
    val db = opts.url.database
    for
      d <- driver()
      _ <- Ledger.ensure(d, db)
      files <- Migration.listMigrationFiles(opts.migrationsDir)
      applied <- Ledger.appliedVersions(d, db)
      _ <- applied.lastOption match
        case None =>
          if opts.verbose then System.err.println("Rollback: nothing to roll back")
          Right(())
        case Some(version) =>
          files.find(_.version == version) match
            case None =>
              Left(Error.msg(s"applied version ${version.value} has no matching migration file"))
            case Some(m) if m.down.isEmpty =>
              Left(Error.msg(s"migration ${m.label} has empty migrate:down"))
            case Some(m) =>
              Runner.applyDown(d, db, m.version, m.down, opts.verbose).map: _ =>
                println(s"Rolled back: ${m.label}")
    yield ()

  def status(quiet: Boolean): Result[Boolean] =
    val db = opts.url.database
    for
      d <- driver()
      exists <- Runner.edge(d.databases().contains(db))
      applied <-
        if exists then Ledger.appliedVersions(d, db).orElse(Right(Nil))
        else Right(Nil)
      files <- Migration.listMigrationFiles(opts.migrationsDir)
      _ <- if opts.strict then Migration.checkStrictOrder(files, applied) else Right(())
      rows = Migration.statusRows(files, applied)
      pending = rows.exists(_._2 == MigrationStatus.Pending)
      _ =
        if !quiet then
          if rows.isEmpty then println("No migrations found.")
          else
            rows.foreach: (m, st) =>
              val mark = st match
                case MigrationStatus.Applied => "[X]"
                case MigrationStatus.Pending => "[ ]"
              println(s"$mark ${m.label}")
    yield pending

  def dump(): Result[Unit] =
    val dbName = opts.url.database
    for
      d <- driver()
      schema <- Runner.edge(d.databases().get(dbName).schema())
      applied <- Ledger.appliedVersions(d, dbName).orElse(Right(Nil))
      files <- Migration.listMigrationFiles(opts.migrationsDir)
      _ <- Runner.edge:
        Option(opts.schemaFile.getParent).foreach(Files.createDirectories(_))
        val header = StringBuilder("-- Schema dumped by tqlmate\n-- Applied migrations:\n")
        if applied.isEmpty then header.append("--   (none)\n")
        else
          applied.foreach: v =>
            files.find(_.version == v) match
              case Some(m) => header.append(s"--   ${m.label}\n")
              case None    => header.append(s"--   ${v.value}\n")
        header.append('\n')
        header.append(schema.trim)
        header.append('\n')
        Files.writeString(opts.schemaFile, header.toString)
        println(s"Wrote: ${opts.schemaFile}")
    yield ()

  def load(): Result[Unit] =
    for
      text <- Runner.edge(Files.readString(opts.schemaFile))
      body = Runner.stripDumpHeader(text)
      _ <- if body.trim.isEmpty then Left(Error.msg("schema file is empty")) else Right(())
      d <- driver()
      db = opts.url.database
      _ <- Runner.edge:
        if !d.databases().contains(db) then d.databases().create(db)
      _ <- Ledger.schemaQueries(d, db, List(body))
      _ = println(s"Loaded: ${opts.schemaFile}")
    yield ()

  def waitReady(timeout: FiniteDuration): Result[Unit] =
    Runner.waitForServer(opts.url, timeout, opts.verbose)

  def up(): Result[Unit] =
    create().flatMap(_ => migrate())

object Runner:
  def resolveUrl(cliUrl: Option[String], dotenv: Map[String, String] = Map.empty): Result[TypeDbUrl] =
    def env(key: String): Option[String] =
      sys.env.get(key).orElse(dotenv.get(key))
    val raw =
      cliUrl
        .orElse(env("TYPEDB_URL"))
        .orElse(env("DATABASE_URL"))
        .getOrElse("typedb://admin:password@localhost:1729/typedb")
    TypeDbUrl.parse(raw)

  def defaultMigrationsDir: Path =
    Path.of(sys.env.getOrElse("TQLMATE_MIGRATIONS_DIR", "db/migrations"))

  def defaultSchemaFile: Path =
    Path.of(sys.env.getOrElse("TQLMATE_SCHEMA_FILE", "db/schema.tql"))

  private[tqlmate] def openDriver(url: TypeDbUrl): Result[Driver] =
    edge:
      val tls =
        if url.tls then DriverTlsConfig.enabledWithNativeRootCA()
        else DriverTlsConfig.disabled()
      TypeDB.driver(url.address, Credentials(url.username, url.password), DriverOptions(tls))

  private[tqlmate] def waitForServer(
    url: TypeDbUrl,
    timeout: FiniteDuration,
    verbose: Boolean
  ): Result[Unit] =
    val deadline = Instant.now().plusMillis(timeout.toMillis)
    def loop(): Result[Unit] =
      openDriver(url) match
        case Right(d) =>
          d.close()
          if verbose then System.err.println(s"TypeDB available at ${url.address}")
          Right(())
        case Left(e) =>
          if Instant.now().isAfter(deadline) then
            Left(Error.msg(s"timed out waiting for TypeDB at ${url.address}: ${e.message}"))
          else
            if verbose then System.err.println(s"waiting for TypeDB at ${url.address} (${e.message})")
            Thread.sleep(500)
            loop()
    loop()

  private[tqlmate] def applyUp(
    driver: Driver,
    database: String,
    version: Version,
    up: String,
    verbose: Boolean
  ): Result[Unit] =
    if up.trim.isEmpty then Left(Error.msg(s"migration ${version.value} has empty migrate:up"))
    else
      val appliedAt = DateTimeFormatter
        .ofPattern("yyyy-MM-dd'T'HH:mm:ss")
        .withZone(ZoneOffset.UTC)
        .format(Instant.now().truncatedTo(ChronoUnit.SECONDS))
      val insert = Ledger.recordInsert(version, appliedAt)
      if verbose then System.err.println(s"-> up ${version.value}")
      Ledger.schemaQueries(driver, database, List(up, insert))

  private[tqlmate] def applyDown(
    driver: Driver,
    database: String,
    version: Version,
    down: String,
    verbose: Boolean
  ): Result[Unit] =
    val delete = Ledger.recordDelete(version)
    if verbose then System.err.println(s"-> down ${version.value}")
    Ledger.schemaQueries(driver, database, List(down, delete))

  private[tqlmate] def stripDumpHeader(text: String): String =
    text.linesIterator
      .dropWhile: l =>
        val t = l.trim
        t.isEmpty || t.startsWith("--")
      .mkString("\n")

  private[tqlmate] def edge[A](body: => A): Result[A] =
    try Right(body)
    catch case e: Exception => Left(Error.fromThrowable(e))

  extension [A](r: Result[A])
    private[tqlmate] def orElse(fallback: Result[A]): Result[A] =
      r match
        case Left(_)  => fallback
        case Right(a) => Right(a)

File: src/runner.rs
use std::path::PathBuf;
use std::time::Duration;

use typedb_driver::{Addresses, Credentials, DriverOptions, DriverTlsConfig, TypeDBDriver};

use crate::ledger::{self, dump_header, schema_queries, strip_dump_header};
use crate::migration::{
    self, check_strict_order, list_migration_files, new_migration_path, status_rows,
    MigrationStatus, Version,
};
use crate::url::TypeDbUrl;
use crate::{Error, Result};

#[derive(Debug, Clone)]
pub struct Opts {
    pub url: TypeDbUrl,
    pub migrations_dir: PathBuf,
    pub schema_file: PathBuf,
    pub strict: bool,
    pub verbose: bool,
    pub wait_timeout: Option<Duration>,
}

pub struct Runner {
    opts: Opts,
    driver: Option<TypeDBDriver>,
}

impl Runner {
    pub fn new(opts: Opts) -> Self {
        Self { opts, driver: None }
    }

    async fn connect(&mut self) -> Result<&TypeDBDriver> {
        if self.driver.is_none() {
            if let Some(timeout) = self.opts.wait_timeout {
                wait_for_server(&self.opts.url, timeout, self.opts.verbose).await?;
            }
            self.driver = Some(open_driver(&self.opts.url).await?);
        }
        self.driver
            .as_ref()
            .ok_or_else(|| Error::msg("internal: driver missing after connect"))
    }

    pub async fn create(&mut self) -> Result<()> {
        let name = self.opts.url.database.clone();
        let verbose = self.opts.verbose;
        let driver = self.connect().await?;
        if driver.databases().contains(name.clone()).await? {
            if verbose {
                eprintln!("database already exists: {name}");
            }
            return Ok(());
        }
        driver.databases().create(name.clone()).await?;
        println!("Created: {name}");
        Ok(())
    }

    pub async fn drop(&mut self) -> Result<()> {
        let name = self.opts.url.database.clone();
        let verbose = self.opts.verbose;
        let driver = self.connect().await?;
        if !driver.databases().contains(name.clone()).await? {
            if verbose {
                eprintln!("database does not exist: {name}");
            }
            return Ok(());
        }
        let db = driver.databases().get(name.clone()).await?;
        db.delete().await?;
        println!("Dropped: {name}");
        Ok(())
    }

    pub async fn new_migration(&self, name: &str) -> Result<PathBuf> {
        std::fs::create_dir_all(&self.opts.migrations_dir)?;
        let path = new_migration_path(&self.opts.migrations_dir, name);
        std::fs::write(&path, migration::migration_template())?;
        println!("Created: {}", path.display());
        Ok(path)
    }

    pub async fn migrate(&mut self) -> Result<()> {
        let db = self.opts.url.database.clone();
        let dir = self.opts.migrations_dir.clone();
        let strict = self.opts.strict;
        let verbose = self.opts.verbose;

        let driver = self.connect().await?;
        if !driver.databases().contains(db.clone()).await? {
            return Err(Error::DatabaseMissing(db));
        }
        ledger::ensure(driver, &db).await?;
        let files = list_migration_files(&dir)?;
        let applied = ledger::applied_versions(driver, &db).await?;
        if strict {
            check_strict_order(&files, &applied)?;
        }
        let pending: Vec<_> = files
            .into_iter()
            .filter(|f| !applied.iter().any(|v| v == &f.version))
            .collect();
        if pending.is_empty() {
            if verbose {
                eprintln!("Migrations: nothing to apply");
            }
            return Ok(());
        }
        for m in pending {
            apply_up(driver, &db, &m.version, &m.up, verbose).await?;
            println!("Applied: {}", m.label());
        }
        Ok(())
    }

    pub async fn rollback(&mut self) -> Result<()> {
        let db = self.opts.url.database.clone();
        let dir = self.opts.migrations_dir.clone();
        let verbose = self.opts.verbose;

        let driver = self.connect().await?;
        ledger::ensure(driver, &db).await?;
        let files = list_migration_files(&dir)?;
        let mut applied = ledger::applied_versions(driver, &db).await?;
        let Some(version) = applied.pop() else {
            if verbose {
                eprintln!("Rollback: nothing to roll back");
            }
            return Ok(());
        };
        let m = files
            .iter()
            .find(|f| f.version == version)
            .ok_or_else(|| Error::MissingMigrationFile(version.clone()))?;
        if m.down.is_empty() {
            return Err(Error::EmptyDown {
                version: m.version.clone(),
                name: m.name.clone(),
            });
        }
        apply_down(driver, &db, &m.version, &m.down, verbose).await?;
        println!("Rolled back: {}", m.label());
        Ok(())
    }

    pub async fn status(&mut self, quiet: bool) -> Result<bool> {
        let db = self.opts.url.database.clone();
        let dir = self.opts.migrations_dir.clone();
        let strict = self.opts.strict;

        let driver = self.connect().await?;
        let applied = if driver.databases().contains(db.clone()).await? {
            ledger::applied_versions(driver, &db)
                .await
                .unwrap_or_default()
        } else {
            Vec::new()
        };
        let files = list_migration_files(&dir)?;
        if strict {
            check_strict_order(&files, &applied)?;
        }
        let rows = status_rows(&files, &applied);
        let pending = rows.iter().any(|(_, s)| *s == MigrationStatus::Pending);
        if !quiet {
            for (m, status) in &rows {
                let mark = match status {
                    MigrationStatus::Applied => "[X]",
                    MigrationStatus::Pending => "[ ]",
                };
                println!("{mark} {}", m.label());
            }
            if rows.is_empty() {
                println!("No migrations found.");
            }
        }
        Ok(pending)
    }

    pub async fn dump(&mut self) -> Result<()> {
        let db_name = self.opts.url.database.clone();
        let schema_file = self.opts.schema_file.clone();
        let dir = self.opts.migrations_dir.clone();

        let driver = self.connect().await?;
        let db = driver.databases().get(db_name.clone()).await?;
        let schema = db.schema().await?;
        let applied = ledger::applied_versions(driver, &db_name)
            .await
            .unwrap_or_default();
        let files = list_migration_files(&dir)?;

        if let Some(parent) = schema_file.parent() {
            std::fs::create_dir_all(parent)?;
        }
        let mut out = dump_header(&applied, &files);
        out.push_str(schema.trim());
        out.push('\n');
        std::fs::write(&schema_file, out)?;
        println!("Wrote: {}", schema_file.display());
        Ok(())
    }

    pub async fn load(&mut self) -> Result<()> {
        let schema_file = self.opts.schema_file.clone();
        let db = self.opts.url.database.clone();
        let text = std::fs::read_to_string(&schema_file)?;
        let body = strip_dump_header(&text);
        if body.trim().is_empty() {
            return Err(Error::EmptySchema);
        }

        let driver = self.connect().await?;
        if !driver.databases().contains(db.clone()).await? {
            driver.databases().create(db.clone()).await?;
        }
        schema_queries(driver, &db, &[body.as_str()]).await?;
        println!("Loaded: {}", schema_file.display());
        Ok(())
    }

    pub async fn wait(&mut self, timeout: Duration) -> Result<()> {
        wait_for_server(&self.opts.url, timeout, self.opts.verbose).await
    }

    pub async fn up(&mut self) -> Result<()> {
        self.create().await?;
        self.migrate().await?;
        Ok(())
    }
}

async fn open_driver(url: &TypeDbUrl) -> Result<TypeDBDriver> {
    let addresses = Addresses::try_from_address_str(url.address())?;
    let credentials = Credentials::new(&url.username, &url.password);
    let tls = if url.tls {
        DriverTlsConfig::enabled_with_native_root_ca()
    } else {
        DriverTlsConfig::disabled()
    };
    Ok(TypeDBDriver::new(addresses, credentials, DriverOptions::new(tls)).await?)
}

async fn wait_for_server(url: &TypeDbUrl, timeout: Duration, verbose: bool) -> Result<()> {
    let start = std::time::Instant::now();
    loop {
        match open_driver(url).await {
            Ok(_) => {
                if verbose {
                    eprintln!("TypeDB available at {}", url.address());
                }
                return Ok(());
            }
            Err(e) => {
                if start.elapsed() >= timeout {
                    return Err(Error::WaitTimeout {
                        address: url.address(),
                        cause: e.to_string(),
                    });
                }
                if verbose {
                    eprintln!("waiting for TypeDB at {} ({e})", url.address());
                }
                tokio::time::sleep(Duration::from_millis(500)).await;
            }
        }
    }
}

async fn apply_up(
    driver: &TypeDBDriver,
    database: &str,
    version: &Version,
    up: &str,
    verbose: bool,
) -> Result<()> {
    if up.trim().is_empty() {
        return Err(Error::EmptyUp(version.clone()));
    }
    let applied_at = chrono::Utc::now().format("%Y-%m-%dT%H:%M:%S").to_string();
    let insert = ledger::record_insert(version, &applied_at);
    if verbose {
        eprintln!("-> up {version}");
    }
    schema_queries(driver, database, &[up, insert.as_str()]).await
}

async fn apply_down(
    driver: &TypeDBDriver,
    database: &str,
    version: &Version,
    down: &str,
    verbose: bool,
) -> Result<()> {
    let delete = ledger::record_delete(version);
    if verbose {
        eprintln!("-> down {version}");
    }
    schema_queries(driver, database, &[down, delete.as_str()]).await
}

pub fn resolve_url(cli_url: Option<&str>) -> Result<TypeDbUrl> {
    let raw = cli_url
        .map(str::to_string)
        .or_else(|| std::env::var("TYPEDB_URL").ok())
        .or_else(|| std::env::var("DATABASE_URL").ok())
        .unwrap_or_else(|| "typedb://admin:password@localhost:1729/typedb".into());
    TypeDbUrl::parse(&raw)
}

pub fn default_migrations_dir() -> PathBuf {
    PathBuf::from(
        std::env::var("TQLMATE_MIGRATIONS_DIR").unwrap_or_else(|_| "db/migrations".into()),
    )
}

pub fn default_schema_file() -> PathBuf {
    PathBuf::from(std::env::var("TQLMATE_SCHEMA_FILE").unwrap_or_else(|_| "db/schema.tql".into()))
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn resolve_url_prefers_argument() {
        let u = resolve_url(Some("typedb://a:b@h:9/argdb")).unwrap();
        assert_eq!(u.database, "argdb");
        assert_eq!(u.port, 9);
    }

    #[test]
    fn default_paths() {
        // Unset-sensitive; just ensure they return something non-empty.
        assert!(!default_migrations_dir().as_os_str().is_empty());
        assert!(!default_schema_file().as_os_str().is_empty());
    }
}
